### ISSUE TITLE
1.11.2 (Beta 81 Testing) Cascading Terrain Lag

### ISSUE THREAD

**Issue by Endmateria** [micdoodle8/Galacticraft#2689]
Getting a "cascading terrain lag" message in the server log.  Quark is trying to deal with this too at the moment... there's a bunch of useful links and suggestions on how to solve it in his issue thread... https://github.com/Vazkii/Quark/issues/412

**Comment by radfast** [comment]
Please post the actual server log so I can see what's causing this in Galacticraft.

**Comment by Endmateria** [comment]
Well I'm having issues playing around with the StructurePro mod at the moment.  I'm troubleshooting that mod with him at the moment.  Here is that crash report... https://gist.github.com/Endmateria/19b081faa7ec6bfe82cd9c2203b111f4#file-crash-2017-05-06_16-24-58-server-txt 
I think the world generators are fighting with each other.  Regardless, if you've never gotten complaints of cascading terrain lag before, then just disregard this issue for a moment while I try to reproduce it without StructPro installed.  And sorry, didn't mean to leave you hanging with a half-assed bug report, just doing heavy multitasking over here.

**Comment by radfast** [comment]
I'm only seeing Galacticraft referenced once in that.

I think most mod developers don't really understand how Minecraft oregen works.

It's roughly like this.

Stage 1:  chunks are generated with basic bedrock, rock, dirt and water, according to the heightmap

Stage 2:  already generated chunks are "populated", but the vanilla Minecraft chunkprovider algorithm will only populate a chunk if at least 3 neighbors already populated, maybe it also depends **which** neighbors those are, you can study the vanilla code.  (One exception is underwater temples which have a special rule.)

Therefore, a chunk can be generated but flagged not "populated".

Getting and setting blockstates in the world causes attempts to load those chunks, which causes generation of those chunks if not already generated.

Vanilla populate code - for example vanilla oregen - can use getting and setting to load chunks either side of the current chunk, so x and z in the range { -1, 0, 1 } from the current chunk's position.  This is obviously necessary to build complete ore pockets or complete trees.  But the oregen and tree code is carefully limited in the range it can cover, if you look at all the vanilla code you'll see it never strays more than around +/-20 blocks from the center of the current chunk.  Therefore, it never triggers chunkgen outside the range  { -1, 0, 1 }.

Because of the rule about only populating a chunk which has 3 populated neighbors already, this makes it impossible for vanilla worldgen to run away. 

If any mod has code which gets or sets outside that tight range, it will cause runaway worldgen, simple as that.

I see ```Tinker's Construct``` slime islands mentioned in that code, as well as vanilla emerald gen.  There's something about vanilla emerald gen which breaks these rules slightly, I don't remember it.

Therefore I doubt that this is a Galacticraft issue, but if factors are pointing to it being us, please explain and I'll look at it some more.

**Comment by Endmateria** [comment]
Yeah, I'm getting cascading terrain messages from a lot of other mods but your mod isn't popping up anymore now that I've removed StructPro.  Thanks for the detailed response/insight though, I'll pass your info along to some of the other mod authors where applicable.

**Comment by TensorTime** [comment]
I am still having pretty major cascade generation and the only culprit left in my pack is galacticraft.  Recurrent Complex I just yanked but it was causing it as well.  Now the only cascade logs I get are galacticraft related.  Here is a screenshot of the runaway generation it is causing by automapping with journey map.  Bear in mind I did not move but just rotated my view to cause this generation.
![2017-06-21_16 18 19](https://user-images.githubusercontent.com/28228945/27409349-4030b72c-569e-11e7-90dd-a13d93154f53.png)

**Comment by TensorTime** [comment]
The pack is available on curse https://minecraft.curseforge.com/projects/my-own-sink and I have an open issue.  I'm currently on 1.11.2-4.0.0-108.  

Thanks!

**Comment by TensorTime** [comment]
Here is a snip from the game output:

```
[16:14:21] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 2595ms behind, skipping 51 tick(s)
[16:14:21] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-25, -13  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:25] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-28, -42  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:25] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-29, -42  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:27] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-12, -44  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:27] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-13, -44  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:27] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-33, -19  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:35] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-35, -20  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:37] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-33, -17  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:40] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-8, -42  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:42] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-34, -38  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:45] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-30, -43  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:46] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-35, -17  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:48] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-34, -16  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
[16:14:48] [Server thread/WARN] [FML]: Galacticraft Core loaded a new chunk (-35, -16  Dimension: 0) during chunk population, causing cascading worldgen lag. Please report this to the mod's issue tracker. This log can be disabled in the Forge config.
```

**Comment by Endmateria** [comment]
@TensorTime  I've been told that the best test for cascading worldgen is to just to creative fly in a straight line (as opposed to just messing with the rendering distance).  Regardless, there's waaaay too many mods in that modpack.  I got 84 mods installed, and I got zero hits for worldgen lag from galacticraft when flying in a straight line for 5 minutes (and I'm using the same version of Galacticraft as you).  I recommend ditching that overkill modpack.

**Comment by TensorTime** [comment]
Thanks for the advice but I've cut every mod that causes world gen cascade.  Galacticraft is the obvious culprit.  I have plenty of people playing the pack and no complaints it is "oversized" considering many are not content related.  I guess we should never have had Sky Factory 3,  it's got too many mods.  

Seriously, how about helpful comments instead of telling people to change their pack because it's too big.

I didn't mess with render distance,  I was in creative and all I needed to do to see the cascade was literaly rotate my character.

I've done the flying around in creative as well,  it's the reason certain mods are not in the pack any more and I switched for something alternative.  Recurrent Complex comes to mind as one of the worst offenders.

The only thing causing the cascade at this point is oregen and the mod doing that generation is galacticraft.  Now it could be because of some other mod aggravating how galacticraft is handling it's oregen,  but that does not mean some other mod is the cause.

**Comment by Endmateria** [comment]
@TensorTime  Well, the general rule is 100 mods, which is considered "heavily modded", and that pack has what, 200+? (I'm not counting them)  Honesty, I don't care how many mods you play with, not my game, though in my experience, the game starts feeling clunky (heavy) at somewhere between 65 and 100 mods (depending on the mods used).  And yes, modpacks can be too big, though it's relatively subjective i.e., dependent on how sluggish and potentially unstable you want your game to be.  It's a question of how far you're willing to push the game engine, at the expense of quality.  Regardless, the only reason I was getting worldgen lag from Galacticraft in the OP was because I was using Structpro, a mod that works as both a structure spawner and a pregenerator.  That mod caused a lot of other mods to display worldgen lag errors even though their worldgen was coded properly.  If you can't reproduce the worldgen lag with only Galacticraft installed, then Galacticraft is not so "OBVIOUSLY" the culprit.  It's poor logic to ignore the multitude of other variables (mods) that can be affecting Galacticraft's worldgen, but understandable if you haven't dealt with the nuances of worldgen lag previously.  You want a solution?  Try it with only Galacticraft installed.  If you get no worldgen hits then put all your mods back on, and then remove 30 mods at a time in batches until the worldgen lag goes away.  When it goes away, then you'll know that the conflicting mod was in the last batch you removed.... and slowly work it down from there until you isolate the conflicting mod/mods.  You know, spend hours doing what the modpack author should have done to begin with.

**Comment by TensorTime** [comment]
Thanks for your unhelpful comments again. Oh I've spent more than hours on the pack and it's been in development for two months.

Seriously if you are not a dev on this mod, stop making rude and unhelpful suggestions. Or am I wrong and you are a mod dev for galacticraft?

**Comment by Endmateria** [comment]
Wow, someone's got a bug up their ass.  I take it that you have difficulty with constructive criticism?  I didn't think that my comments were that unhelpful.  I pointed out that other mods can influence a worldgen lag error, that you can deduce the cause of the problem by trial and error, and that your modpack was too big (which it is) hahaha.  I also didn't realize that you were the modpack author, though I can't say that I would have sugarcoated my responses in light of that fact.  So you ignore my feedback on how to isolate the problem?  What do you expect the Galacticraft author to waste hours of his time trying to solve a worldgen lag error caused by your mod setup?  Personally, I would start the test by removing Doomlike Dungeons, Abyssalcraft, OceanFloor, and Largeveins (those have a lot of big gen).  Extra Planets has been incredibly unstable lately in my experience (causing timeout crashes).  Solar Flux Reborn and Enviromental Tech do roughly the same thing.  Nobody is going to use Nuclearcraft for power when they have Env Tech and Extreme Reactors.  Everything that Buildcraft and Tech Reborn do can be handled by the other mods on your list.  Scannable and Advanced finder compass do the same thing.  Mystical Agriculture makes Gendustry obsolete.  Also, CoFH core has a config for flattening bedrock, so you can remove the better bedrock gen mod.  Honestly, I'm surprised that you didn't put both Waila and the One Probe on the list.  And no I'm not a Galacticraft dev, just a guy who wasted his time trying to help you when you got no response from the dev.  You're welcome douche hahaha.

**Comment by Endmateria** [comment]
https://www.youtube.com/watch?v=F-mju_gW3c8

**Comment by TensorTime** [comment]
I apologize to the Galacticraft devs for this thread turning into the troll job it apparently has.  I will continue on my end to try and lock down what's happening in the oregen and try and see if I can reproduce the issue or make it go away by tweaking configs of other mods that have subsurface generation of some sort.

In the meantime let me know if there is anything I can do to help you or any testing you would like me to do.

Thanks

**Comment by TensorTime** [comment]
I created a fresh profile in the twitch app.   Set it to 1.11.2 with the latest forge (13.20.0.2315 at this point) and added absolutely nothing but the 3 mods from galacticraft.  I've attached the manifest from the profile which basically has nothing else but the galacticraft mods and generated configs in the overrides folder tree.

https://gist.github.com/TensorTime/7527ab82747339900138dd8361adb4bb

Quite clearly,  as I said in the beginning,  nothing else is causing this cascade.  Nothing else is installed to be aggravating it other than forge and what it might put in there.

Just for context for the logs.  I created a new creative world,  hopped into flight mode and went in a straight line at a decent speed for about 20 seconds max then logged off and closed the game.  So the complete cycle of events should be captured there.

Let me know what else I can do to help.
Here is the manifest:
```
{
  "minecraft": {
    "version": "1.11.2",
    "modLoaders": [
      {
        "id": "forge-13.20.0.2315",
        "primary": true
      }
    ]
  },
  "manifestType": "minecraftModpack",
  "manifestVersion": 1,
  "name": "GC Test",
  "version": "1.0",
  "author": "tensortime",
  "files": [],
  "overrides": "overrides"
}
```

**Comment by Endmateria** [comment]
Did you try deleting the Galacticraft configs?  The authors stated in their beta thread that the configs should ideally be wiped with each new update.  It's worth a shot if you haven't tried that yet.

**Comment by TensorTime** [comment]
I created a fresh profile.  There were no configs except the fresh generated ones.

**Comment by TensorTime** [comment]
Are you running a stand alone server and connecting to it?  Because I'm using the built in server and they don't function identically.  That's why you need to test your packs separately on servers vs stand alone clients (if you intend it to run in both).

**Comment by TensorTime** [comment]
Also what version of Forge are you on?

**Comment by Endmateria** [comment]
Forge 2315
GC 1.11.2-4.0.0.109

This is way I do it, which I assume is the standard way i.e., I extract the minecraft server from the forge 2315 universal installer file and then create a batch file to launch the forge 2315.jar.  https://www.youtube.com/watch?v=vmUHnGo0GoI
I can't say I'm very educated on other server setups, but I've always used this setup for the small group of people that I play with.

**Comment by TensorTime** [comment]
Yeah I've made server packs for modpacks that don't have them that way. I'm
not running this in a server and I don't expect many people would because
of it's size.  It's intended to be a single player pack but some folks
might convert it to a server since I won't be providing a server pack.

So I wonder if then it has something to do with that fact?  It would be one
of the strangest cascades I've seen (and I've removed at least a dozen
cases by either fixes or config changes).  Because as far as I can tell the
only difference if you tried what I did would be running forge as a server
vs. client install.

Also if you think I didn't read some of your messages, I did.  You did have
some helpful suggestions (like getting rid of BBG and using the CoFH
config,  didn't know that was there).

**Comment by Endmateria** [comment]
I don't know if there's a relationship here, but that Structpro mod I was using messed with the tickrate a lot to do the pregeneration, and that's when I was getting all kinds of faulty worldgen lag reports.  I found this mod last night, you could give it a try just for kicks.  I know when the tickrate drops, that entities and worldgen can't keep up... this mod tries to stabilize that a little bit.  Don't know if it'll help, just trying to think of ways to troubleshoot this with you.  https://minecraft.curseforge.com/projects/tick-dynamic/files

**Comment by Endmateria** [comment]
And yeah, I've never seen that many worldgen lag hits so fast.  I should lauch a single player game with my setup and see if I get the same kind of log output.  I'll actually do that now.  I'll ask the dumb, but obvious question... did you double check to see if you gave your client profile enough ram?

**Comment by TensorTime** [comment]
8 gig on a 32 gig machine yep.

**Comment by TensorTime** [comment]
I'll try it out,  worth a try at least.  The one thing I've noticed through
is I'm not actually getting true "lag" as a result when it's just
galacticraft spewing like that so I don't think it's having a major hit on
TPS (though it's hard to watch both simultaneously).

**Comment by Endmateria** [comment]
Well if I can duplicate it, then you might have a legitimate grievance on your hands (testing now)... though it's kinda bizarre that there'd be a ton of worldgen lag in a client instance, and none in a server instance.

**Comment by Endmateria** [comment]
This was my latest log output, launching a new world with all my mods still installed, client side only.
https://gist.github.com/Endmateria/9e9c4b679fdd4d1961a99afccb12cd04#file-gistfile1-txt  You might want to try moving your mods folder and then deleting your entire minecraft folder (except for the launcher), and reinstalling both Minecraft and Forge.  I've had bizarre problems like this in the past that were only fixed by a clean install.

**Comment by TensorTime** [comment]
Fair enough,  I'll refresh Twitch's common installs and reinstall the small test pack and my pack and see what happens.

**Comment by Endmateria** [comment]
Then again, I still don't think I'm launching the same way you are.  I didn't use Twitch for this test, just the standard game launcher.  I know nothing about Twitch, or how to launch using it.  I assume you're using it to livestream the game, if so then it's over my head hahaha

**Comment by TensorTime** [comment]
Twitch replaced the curse app after they bought Curse.  So it's the old Curse gaming launcher renamed to the Twitch Desktop App :)

Though yes I do stream on occasion lol

**Comment by TensorTime** [comment]
Yea still seeing it:

Setup:

1. Brand new Twitch Desktop App install.
2. One profile for Minecraft 1.11.2 Forge 2315.
3. Nothing but the 3 Galacticraft mods installed.

Gist snippet of log output:
https://gist.github.com/TensorTime/878ab83f66d5e3adc0ba068f14587656

:(

**Comment by TensorTime** [comment]
Going to try a plain forge client install and see what happens

**Comment by Endmateria** [comment]
Yeah, I'm looking at it now, looks interesting, though  I'm gonna have to deduce that the Twitch client may be a significant part of the issue for whatever reason.

**Comment by TensorTime** [comment]
Yeah but if that is the case, it's the most common launcher used for modded minecraft so that could be a bigger issue than this lol.

**Comment by TensorTime** [comment]
Not Twitch.

Ok this is from a vanilla launcher,  using the same 1.11.2-13.20.0.2315 installed with nothing but the 3 Galacticraft mods in the mods folder.  Same cascade problem.

https://gist.github.com/TensorTime/e2879e55139d703a333aef80d7bf2642

**Comment by Endmateria** [comment]
WTF hahaha

**Comment by TensorTime** [comment]
Just out of curiosity, what java are you using?  Perhaps it's something more fundamental like a java version difference.  Currently I'm running 1.8.0.131 64-bit.

**Comment by Endmateria** [comment]
Hahaha, I was just gonna say that.  You got Java Runtime Environments installed (JRE)?  Probably not the issue, but I've heard it helps.

Look at your last log, this seems off...  Java Version: 1.8.0_25, Oracle Corporation
The one before that says... Java Version: 1.8.0_131, Oracle Corporation

Try a clean install of Java and throw on JRE while you're at it.

**Comment by TensorTime** [comment]
Yeah that's the native launcher it uses it's own very old version of Java.

**Comment by radfast** [comment]
Thanks for this guys, if there is cascading worldgen we obviously want to fix it.  It's a bit strange that nobody else has reported this, and generally we would see it in our own testing too.

@TensorTime  please can you give us:
* your world seed for that world (your latest log at https://gist.github.com/TensorTime/e2879e55139d703a333aef80d7bf2642)
* copy of your ```options.txt``` in the .minecraft folder

Any just to be clear, this latest run was in newly created single player world, which had never had any other mods used?

**Comment by Endmateria** [comment]
Ahhh, now I'm seeing it (I was looking at the latest log, rather than the fml client log).  I removed all mods except for the 3 GC ones, deleted all old configs and launched a fresh single player world.

See lines 762-794
https://gist.github.com/Endmateria/ae2f06d438190f1894a32bf91f36c75d#file-fml-client-latest-log

**Comment by TensorTime** [comment]
[options.txt](https://github.com/micdoodle8/Galacticraft/files/1095771/options.txt)

**Comment by Endmateria** [comment]
@TensorTime  Yeah, I saved that, don't feel like losing all my keybinds again hahaha

**Comment by Endmateria** [comment]
https://gist.github.com/Endmateria/fbb92e920567065f4c5fcb95bc74ddcc#file-options-txt

**Comment by TheQuadShot** [comment]
For what its worth... This has probably been around for a long time, but Forge didn't log it untill 1.11.x (https://github.com/MinecraftForge/MinecraftForge/pull/3756) and the most common way to "solve" the issue is:

"offset the gen position by 8 on the X/Z axes. If you don't, you could run into cascading chunk gen problems." (http://www.minecraftforge.net/forum/topic/53563-111-world-generation/)

**Comment by TensorTime** [comment]
It actually goes away when I turn off oil, but not when I turn off ores, so I'm suspecting it still is the oil gen.  I went through and individually tested each ore and the oil and the only runaway is with the oil as far as I can tell.

**Comment by radfast** [comment]
The Refinery accepts any Forge fluid whose name is "oil" - but not other fluids like "heavy oil" or "olive oil".

Immersive Petroleum's Crude Oil is already a fluid whose name is "oil", so it should just work - did you put some into the Galacticraft Refinery to try it?

**Comment by Endmateria** [comment]
I know that the GC Refinery works with Actually Additions oil.  Actually, because AA loads first in the mod list, all GC oil gen shows up as Actually Additions oil in the one probe/waila description menu.  Not a problem given that the oil looks and works the same.

**Comment by radfast** [comment]
@TensorTime  thank you for your tests on the worldgen.

The issue with oil gen should be fixed in build 114 onwards, please re-open if you see any more "cascading world gen" logs involving Galacticraft.  I've run a few test worlds with different seeds in 1.11.2 and I'm not seeing it now - it does still happen (rarely) from vanilla minecraft, and of course it might be possible with other mods.

**Comment by TensorTime** [comment]
Thank you!! Working like a charm.
